### PR TITLE
Conditional internal migrations steps

### PR DESCRIPTION
**Pull request by Voronenko**
Possibility to have internal migration steps that depend on database state

### CODE REVIEW

File: lib/Doctrine/DBAL/Migrations/AbstractMigration.php [hunk 1]
```diff
@@ -173,6 +173,11 @@ protected function addSql($sql, array $params = [], array $types = [])
         $this->version->addSql($sql, $params, $types);
     }
 
+    protected function checkResultExists($sql, array $params = array(), array $types = array())
+    {
+      return $this->version->checkResultExists($sql, $params, $types);
```
Comment (stof) [inline lib/Doctrine/DBAL/Migrations/AbstractMigration.php]: wrong indentation

File: lib/Doctrine/DBAL/Migrations/AbstractMigration.php [hunk 2]
```diff
@@ -173,6 +173,11 @@ protected function addSql($sql, array $params = [], array $types = [])
         $this->version->addSql($sql, $params, $types);
     }
 
+    protected function checkResultExists($sql, array $params = array(), array $types = array())
```
Comment (stof) [inline lib/Doctrine/DBAL/Migrations/AbstractMigration.php]: phpdoc should be added

File: lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php [hunk 3]
```diff
@@ -703,6 +703,13 @@ public function getMigrationsToExecute($direction, $to)
         return $versions;
     }
 
+
+    public function setMappingTypes($mapping_types) {
+        foreach($mapping_types as $field=>$type) {
+            $this->connection->getDatabasePlatform()->registerDoctrineTypeMapping($field, $type);
```
Comment (stof) [inline lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php]: changing the configuration of the connection looks wrong to me. the connection is passed as a constructor argument to this class, which means that configuring it is meant to be done outside this class

**Comment by stof** [comment]
@Voronenko you could just check the Schema object passed as argument

File: lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php (hunk 3 above)
Comment (stof) [inline lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php]: and btw, this seems unrelated to the feature you are trying to add here

**Comment by mikeSimonson** [comment]
@Voronenko 

There are 2 unrelated stuff in your migration.
A new method and the type mapping.
Please make 2 different pull request.

The build folder has to stay there for the tests.

I don't think that the new method checkResultExists is interesting enough so that it should be there by default. You can add it in your version of the AbstractMigration without touching the version.

If you want me to merge the type mapping pull request, you need to add tests.

File: lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php (hunk 3 above)
Comment (Voronenko) [inline lib/Doctrine/DBAL/Migrations/Configuration/Configuration.php]: Agree. I will split into two and address comments

**Comment by stof** [comment]
@mikeSimonson IMO, the type mapping stuff is wrong here. The connection is meant to be configured externally.
and if they really want to do such weird config change inside their migration, they can still do it (the connection is accessible). But we should not encourage it by adding a dedicated API to do it
